src/main/kotlin/com/terraformation/backend/tracking/db/ObservationStore.kt

```diff
@@ -809,11 +821,134 @@ class ObservationStore(
               .isEmpty()
 
       if (allPlotsCompleted) {
-        completeObservation(observationId, plantingSiteId)
+        completeObservation(observationId, plantingSiteId, isAdHoc)
       }
     }
   }
 
+  fun insertBiomassDetails(
+      observationId: ObservationId,
+      plotId: MonitoringPlotId,
+      model: NewBiomassDetailsModel
+  ) {
+    requirePermissions { updateObservation(observationId) }
+
+    val (observationType, observationState) =
+        with(OBSERVATION_PLOTS) {
+          dslContext
+              .select(
+                  observations.OBSERVATION_TYPE_ID.asNonNullable(),
+                  observations.STATE_ID.asNonNullable())
+              .from(this)
+              .where(OBSERVATION_ID.eq(observationId))
+              .and(MONITORING_PLOT_ID.eq(plotId))
+              .fetchOne()
+              ?: throw IllegalStateException(
+                  "Plot $plotId is not part of observation $observationId")
+        }
+
+    if (observationState == ObservationState.Completed) {
+      throw IllegalStateException("Observation $observationId is already completed.")
+    }
+
+    if (observationType != ObservationType.BiomassMeasurements) {
+      throw IllegalStateException("Observation $observationId is not a biomass measurement")
+    }
+
+    model.validate()
+
+    dslContext.transaction { _ ->
+      with(OBSERVATION_BIOMASS_DETAILS) {
+        dslContext
+            .insertInto(this)
+            .set(OBSERVATION_ID, observationId)
+            .set(MONITORING_PLOT_ID, plotId)
+            .set(DESCRIPTION, model.description)
+            .set(FOREST_TYPE_ID, model.forestType)
+            .set(SMALL_TREES_COUNT_LOW, model.smallTreeCountRange.first)
+            .set(SMALL_TREES_COUNT_HIGH, model.smallTreeCountRange.second)
+            .set(HERBACEOUS_COVER_PERCENT, model.herbaceousCoverPercent)
+            .set(SOIL_ASSESSMENT, model.soilAssessment)
+            .set(WATER_DEPTH_CM, model.waterDepthCm)
+            .set(SALINITY_PPT, model.salinityPpt)
+            .set(PH, model.ph)
+            .set(TIDE_ID, model.tide)
+            .set(TIDE_TIME, model.tideTime)
+            .execute()
+      }
+
+      val quadratDetailsRecords =
+          model.quadrats.map { (position, details) ->
+            ObservationBiomassQuadratDetailsRecord(
+                observationId, plotId, position, details.description)
+          }
+      dslContext.batchInsert(quadratDetailsRecords).execute()
+
+      val quadratSpeciesRecords =
+          model.quadrats.flatMap { (position, details) ->
+            details.species.map {
+              ObservationBiomassQuadratSpeciesRecord(
+                  observationId,
+                  plotId,
+                  position,
+                  it.speciesId,
+                  it.speciesName,
+                  it.isInvasive,
+                  it.isThreatened,
+                  it.abundancePercent,
```

Comment:
Named parameters would add a little extra safety here since there are, e.g., two boolean arguments in a row and we might not notice if they got swapped by accident. Same with the other `Record` class constructor calls in the function.